File: apps/nest/deploy/domain/interfaces.go
package domain

import (
	"io/fs"
	"time"
)

// FileInfo represents basic file metadata from a remote filesystem.
type FileInfo interface {
	Name() string
	Size() int64
	Mode() fs.FileMode
	ModTime() time.Time
	IsDir() bool
}

// RemoteFS abstracts remote filesystem operations (e.g., SFTP).
type RemoteFS interface {
	// Stat returns file info. Returns error if the file does not exist.
	Stat(path string) (FileInfo, error)
	// MkdirAll creates a directory and all parents.
	MkdirAll(path string) error
	// ReadDir returns child names in a directory (non-recursive).
	ReadDir(path string) ([]string, error)
	// ReadFile reads file contents.
	ReadFile(path string) ([]byte, error)
	// WriteFile writes content to a file, creating it if necessary.
	WriteFile(path string, data []byte) error
	// Remove removes a file or directory recursively.
	Remove(path string) error
	// Rename moves/renames a file or directory.
	Rename(src, dst string) error
	// FileHash computes SHA256 hash of a remote file.
	FileHash(path string) (string, error)
	// FileModTime returns the modification time of a file.
	FileModTime(path string) (time.Time, error)
}

// RemoteExec abstracts remote command execution (e.g., SSH).
type RemoteExec interface {
	// Exec runs a command and returns combined output error.
	Exec(command string) error
	// ExecPipe runs a command with stdout/stderr piped to the console.
	ExecPipe(command string) error
}

// UserPrompter abstracts user interaction for conflict resolution.
type UserPrompter interface {
	// AskConflictAction asks the user how to handle conflicting files.
	// Returns the chosen action and backup suffix (if backup was chosen).
	AskConflictAction(files []string, lang string) (ConflictAction, string, error)
}

// SnapshotRepository abstracts reading/writing snapshot metadata.
type SnapshotRepository interface {
	// Read returns the snapshot from the target directory.
	// Returns nil (no error) if no snapshot exists.
	Read(targetDir string) (*Snapshot, error)
	// Write persists the snapshot to the target directory.
	Write(targetDir string, snapshot *Snapshot) error
}

File: apps/ship/src/deploy/domain/interfaces.ts
import type { ConflictAction } from "./conflict.js";
import type { Snapshot } from "./snapshot.js";

export interface RemoteFileInfo {
  name: string;
  size: number;
  isDir: boolean;
  modTime: Date;
}

/** Remote filesystem operations (SFTP). */
export interface RemoteFS {
  /** null when the path does not exist. */
  stat(path: string): Promise<RemoteFileInfo | null>;
  mkdirAll(path: string): Promise<void>;
  /** Child names (non-recursive), excluding ship metadata. */
  readDir(path: string): Promise<string[]>;
  readFile(path: string): Promise<string>;
  writeFile(path: string, data: string): Promise<void>;
  /** Removes a file or directory recursively; a missing path is not an error. */
  remove(path: string): Promise<void>;
  rename(src: string, dst: string): Promise<void>;
  /** SHA256 of a remote file. */
  fileHash(path: string): Promise<string>;
  fileModTime(path: string): Promise<Date>;
}

/** Remote command execution (SSH). */
export interface RemoteExec {
  exec(command: string): Promise<void>;
  execPipe(command: string): Promise<void>;
}

export interface ConflictDecision {
  action: ConflictAction;
  /** Only meaningful when action === "backup". */
  suffix: string;
}

export interface UserPrompter {
  askConflictAction(files: string[]): Promise<ConflictDecision>;
}

export interface SnapshotRepository {
  /** null when no snapshot exists. */
  read(targetDir: string): Promise<Snapshot | null>;
  write(targetDir: string, snapshot: Snapshot): Promise<void>;
}
